### ISSUE TITLE
firewall: needs the ability to add custom ports

### ISSUE THREAD

**Issue by andreasn** [cockpit-project/cockpit#9048]
This came up during the [usability testing](https://github.com/cockpit-project/cockpit/wiki/Usability-test-Firewall) of the firewall.
Participants expressed a desire to add custom ports. This is most important in these two scenarios:
 * Running a service on a non-standard port, such as running SSH on a different port than 22, because all the script kiddies port scanners are set to look for 22.
 * You have a service that isn't standard/common enough to have a firewall service file

I know this is a planned feature already, so best to create an item to track it.

Mockup by @garrett:
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/firewall/firewall-dialog-add-service.png)

This issue came up in 50% of the tests (3 out of 6 participants)

**Comment by sgallagher** [comment]
Instead of (or in addition to) "custom ports", could we instead have a service editor? It would allow users to name their custom settings and/or make modifications to the standard ones on the system.

Cockpit would drop an appropriate XML file into `/etc/firewalld/services` and it would become available upon a reload of the firewalld service (with `firewall-cmd --reload` or the equivalent DBUS API call).

Then, these custom settings can be managed just like any other service.

Tagging @t-woerner for his thoughts as well.

**Comment by garrett** [comment]
@sgallagher: Yep. That's exactly what this does. Ports aren't handled directly; they're abstracted into services, and the service is auto-named, with the possibility to change it.

**Comment by sgallagher** [comment]
@garrett That doesn't seem to match what the mock-ups show.

**Comment by garrett** [comment]
@sgallagher: It actually is what the mockups show, and is what I wrote in the design doc and talked with the team about as well. I will modify the mockups to make it more obvious, as it isn't obvious enough when they stand alone.

The idea is that everything is a service and adding ports automatically creates a service.

The design document is at https://github.com/cockpit-project/cockpit/wiki/Feature:-Firewall#design-requirements--mockups (with a link to one place where it's specifically mentioned). The mockups are supporting assets for the design doc.

**Comment by garrett** [comment]
@sgallagher: OK, I've added a sentence to clarify that services are created in the `firewall-dialog-add-service.svg` (and `png`) file, so that should help when the mockup is viewed out of context from the design document. It should update in the file above too.

Sorry for the confusion.

**Comment by larskarlitski** [comment]
> Cockpit would drop an appropriate XML file into /etc/firewalld/services and it would become available upon a reload of the firewalld service (with firewall-cmd --reload or the equivalent DBUS API call).

I agree that we need this in general, but I hope that we'll be able to add new services with a [D-Bus call](http://www.firewalld.org/documentation/man-pages/firewalld.dbus.html) as well.

**Comment by sgallagher** [comment]
@garrett I think the part that was confusing to me is that the mock-ups don't show how you would then select and enable/disable the service with the custom name. It wasn't clear that it would just show up in the available services list. It's also not clear that it lets you edit pre-existing services.

@larskarlitski Ah, I didn't realize there was an API for that. Even better!

**Comment by garrett** [comment]
@sgallagher: Yeah, it would immediately add and enable the newly-create service. 

Disabling & enabling would happen in the list, like every other service. (That isn't done yet; we're working on adding features in stages.) It also isn't in the design document, but the usability tests show that we're going to need to be able to toggle enabled/disabled states.

As for editing, we had a conversation that it would be done in a later stage. It could either be edited in a similar dialog or inline. (It hasn't been designed yet.)

**Comment by garrett** [comment]
@Songxuewen: Thanks for reporting #9327. Please follow this issue instead for an update. 

(I hope just @'ing you here might automatically do that?)

**Comment by marusak** [comment]
I am looking into this and, @garrett, I've hit one problem - each new service needs unique identifier (id). The field in design labeled `Name` is more suitable for `Description` or for `Short`, see https://firewalld.org/documentation/man-pages/firewalld.service.html but hardly usable for `id` as it cannot contain spaces and possibly other limitations such as capital letters or some characters... 
Surely we can edit name to fit these rules, but if user would once pick ports `21,22,23,80` we would generate name `ftp,ssh,telnet, etc.` and if picked ports `21,22,23,9090` then we would generate the same name, which can be changed, but it is very bad experience if we generate something and then it is not valid.

Do you have some ideas how we can address this?

**Comment by garrett** [comment]
@marusak: I'm not exactly sure if this matches up to what you want, but a year and a half ago, I did a cross-reference mashup between firewalld services and /etc/services @ https://garrett.github.io/cockpit-mockweb/firewall/services, as a demo to show that it's possible to get more information for custom ports... including names and IDs.

I basically:

1. parsed `/etc/services` and output it as a CSV @ https://github.com/garrett/cockpit-mockweb/blob/master/_data/etc_services.csv

2. exported all the firewalld services at `/usr/lib/firewalld/services` on my machine as JSON @ https://github.com/garrett/cockpit-mockweb/tree/master/_data/firewalld (they're normally XML)

_I chose those CSV and JSON as it's easy to coerce a list of data into CSV and Jekyll supports both JSON and CSV as data stores. _Of course_ Cockpit wouldn't convert them into the intermediary data. This was just done to be able to use the data from Jekyll._ :wink:

3. I then used Liquid templating in Jekyll to crudely smash them together: https://raw.githubusercontent.com/garrett/cockpit-mockweb/master/firewall/services.md

_Liquid only provides a relatively small set of operations, so it had to be a little crude._

---

So, anyway, the above is background to say that we could probably parse `/etc/services`, look up the IDs there, and string them together... and filenames could look like this for the two port examples you gave:

- `custom--ftp-ssh-telnet-http`
- `custom--ftp-ssh-telnet-websm`

The filenames would be more meaningful than just the ports, like:

- `21-22-23-80`
- `21-22-23-9090`

Would this solve the problem?

(If IDs are mapped 1:1 to filenames, I see some have uppercase and dashes. I don't see "ID" in the service file XML... just short, description, port, destination, and sometimes another field or two. So it's probably a pretty safe assumption that ID == {_filename_}.xml)
